### PR TITLE
EDGEPOD-336 : MME-Enhancements to support NSA during attach procedure

### PR DESCRIPTION
**Pull request by clstefi**
This activity addresses the following issues:
1. EDGEPOD-332 NSA MME : S1AP Protocol Layer Support
2. EDGEPOD-333 NSA MME : S11 Gtpv2 Protocol Layer Support
3. EDGEPOD-336 MME-Enhancements to support NSA during attach procedure
4. EDGEPOD-337 NSA MME : NAS Protocol Layer Support

### CODE REVIEW

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

File: include/common/msgType.h [hunk 1]
```diff
@@ -219,7 +221,9 @@ struct handover_cancel_Q_msg {
 struct tauReq_Q_msg {
     int seq_num;
     int enb_fd;
-	int s1ap_enb_ue_id;
+    int s1ap_enb_ue_id;
+    UE_net_capab ue_net_capab;
```
Comment (thakurajayL) [inline include/common/msgType.h:225]: these fields may not be mandatory right ? adding bool variable for presence might be good idea

Comment (clstefi) [reply include/common/msgType.h:225]: Yes, taken care in https://github.com/omec-project/Nucleus/pull/51/commits/d9c789b92c4c58ca8784ef2702f0ec5446ccf359

File: include/s6a/s6a_fd.h [hunk 2]
```diff
@@ -41,8 +41,6 @@
 /*SMS-Only-Indication:             1-------*/
 #define ULR_FLAG_SMS_ONLY_IND (1<<7)
 
-#define EXT_MAX_REQ_BW_DL_AVP_CODE 554
```
Comment (thakurajayL) [inline include/s6a/s6a_fd.h:44]: Why this is removed ?

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 3]
```diff
@@ -327,13 +340,31 @@ ActStatus ActionHandlers::process_ula(SM::ControlBlock& cb)
 	ue_ctxt->setSubscriptionStatus(ula_msg.subscription_status);
 	ue_ctxt->setNetAccessMode(ula_msg.net_access_mode);
 	ue_ctxt->setAccessRestrictionData(ula_msg.access_restriction_data);
-  ue_ctxt->setSubscribedApn(Apn_name(ula_msg.selected_apn));
-  log_msg(LOG_DEBUG, "Selected APN %s \n",ula_msg.selected_apn.val);
+	
+    ue_ctxt->setSubscribedApn(Apn_name(ula_msg.selected_apn));
+    log_msg(LOG_DEBUG, "Selected APN %s \n", ula_msg.selected_apn.val);
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:345]: Get rid of this log. If you have opportunity to do one more change. We really need to cleanup logs.

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 4]
```diff
@@ -327,13 +340,31 @@ ActStatus ActionHandlers::process_ula(SM::ControlBlock& cb)
...
 
+    for (int i = 0; i < ula_msg.supp_features_list.count; i++)
+    {
+        if (ula_msg.supp_features_list.supp_features[i].feature_list_id == 2)
+        {
+            ue_ctxt->setHssFeatList2(
+                    ula_msg.supp_features_list.supp_features[i]);
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:352]: why can't we do this @line 251

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 5]
```diff
@@ -327,13 +340,31 @@ ActStatus ActionHandlers::process_ula(SM::ControlBlock& cb)
...
+	
+	// UE-AMBR received from HSS. Treated as APN-AMBR as well.
+	// Bitrate values beyond 4.2 Gbps will be set in extended AMBR fields.
+	// Extended AMBR fields can store bit rate values upto 4.2 Tbps.
 	struct AMBR ambr;
 	ambr.max_requested_bw_dl = ula_msg.max_requested_bw_dl;
 	ambr.max_requested_bw_ul = ula_msg.max_requested_bw_ul;
-	
+	ambr.ext_max_requested_bw_dl = ula_msg.extended_max_requested_bw_dl;
+	ambr.ext_max_requested_bw_ul = ula_msg.extended_max_requested_bw_ul;
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:366]: do we know if values are always present ?  dcnr is configurable that means if MME does not send dcnr then MME will not get anything from HSS...

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 6]
```diff
@@ -525,14 +556,20 @@ ActStatus ActionHandlers::sec_mode_cmd_to_ue(SM::ControlBlock& cb)
...
 	nas.header.nas_security_param = AUTHREQ_NAS_SECURITY_PARAM;
-	const uint8_t num_nas_elements = SEC_MODE_NO_OF_NAS_IES;
+        uint8_t num_nas_elements = SEC_MODE_NO_OF_NAS_IES;
+	/* UE additional security capability is replayed in sec_mode_command when MME supports the handling of the same
+         * Spec 24.301 v 15.6.0  sec:8.2.20.6*/
+	if(mme_cfg->feature_list.dcnr_support && ue_ctxt->getUeAddSecCapabPres())
+    	{
+            num_nas_elements += 1;
+    	}
 	nas.elements = (nas_pdu_elements *) calloc(num_nas_elements, sizeof(nas_pdu_elements)); // TODO : should i use new ?
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:566]: @asreek  should we avoid using calloc now in Mme cpp files ? Can you please guide us ?

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 7]
```diff
@@ -1028,6 +1094,63 @@ ActStatus ActionHandlers::send_init_ctxt_req_to_ue(SM::ControlBlock& cb)
...
+	if (ue_ctxt->getUeNetCapab().ue_net_capab_m.u.bits.dcnr)
+	{
+		// But the network does not allow DCNR for the UE
+		if (ue_ctxt->getDcnrCapable() == false)
+		{
+			eps_nw_feature_supp_presence = true;
+
+			icr_msg.ho_restrict_list_presence = true;
+			memcpy(&(icr_msg.ho_restrict_list.serving_plmn),
+					&(ue_ctxt->getTai().tai_m.plmn_id), sizeof(struct PLMN));
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:1110]: PLMN structure has multiple extra fields. and that causes PLMN struct to return 6-7 bytes. I guess you just want to copy mcc/mnc. In that case just pass 3 as argument to memcpy.

File: src/mme-app/utils/mmeNasUtils.cpp [hunk 8]
```diff
@@ -569,10 +570,52 @@ int MmeNasUtils::parse_nas_pdu(s1_incoming_msg_data_t* msg_data, unsigned char *
...
+                        msg++; //skipping Element ID
+                        datalen = msg[0];
+                        msg++;
+                        memcpy(&(nas->elements[index].pduElement.ue_add_sec_capab), msg, datalen);
+                        msg += datalen;
+                        index++;
+                        break;
+                    }
+                    default:
+                        msg++;
```
Comment (thakurajayL) [inline src/mme-app/utils/mmeNasUtils.cpp:615]: This does not seem to be correct. Should we not skip the whole AVP/IE else Mme will crash. Thes

File: src/mme-app/actionHandlers/attachActionHandlers.cpp [hunk 9]
```diff
@@ -327,13 +340,31 @@ ActStatus ActionHandlers::process_ula(SM::ControlBlock& cb)
...
+    {
+        if (ula_msg.supp_features_list.supp_features[i].feature_list_id == 2)
+        {
+            ue_ctxt->setHssFeatList2(
+                    ula_msg.supp_features_list.supp_features[i]);
+            break;
+        }
+    }
+        
+	ue_ctxt->setDcnrCapable(MmeCommonUtils::isUeNRCapable(*ue_ctxt));
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/attachActionHandlers.cpp:357]: Why this code ? If ue Is NR capable then DCNR capable ?

File: include/s6a/s6a_fd.h (hunk 2 above)
Comment (clstefi) [reply include/s6a/s6a_fd.h:44]: Previously, we didn't have this AVP support in freediameter. So, this was required.
Now that we have updated the dictionary, we can fetch the AVP code directly from the dictionary.

File: src/mme-app/actionHandlers/attachActionHandlers.cpp (hunk 3 above)
Comment (clstefi) [reply src/mme-app/actionHandlers/attachActionHandlers.cpp]: Done.

File: src/mme-app/actionHandlers/attachActionHandlers.cpp (hunk 5 above)
Comment (clstefi) [reply src/mme-app/actionHandlers/attachActionHandlers.cpp]: Taken care in https://github.com/omec-project/Nucleus/pull/51/commits/d588d536035fba07e1b1ccddd992218318bf99d5

File: src/mme-app/actionHandlers/attachActionHandlers.cpp (hunk 7 above)
Comment (clstefi) [reply src/mme-app/actionHandlers/attachActionHandlers.cpp]: Done.

File: src/mme-app/utils/mmeNasUtils.cpp (hunk 8 above)
Comment (clstefi) [reply src/mme-app/utils/mmeNasUtils.cpp]: Taken care in d588d53

**Comment by clstefi** [comment]
> can you please do changes around memcpy/struct PLMN as suggested

Sure, done.

File: src/mme-app/actionHandlers/attachActionHandlers.cpp (hunk 9 above)
Comment (clstefi) [reply src/mme-app/actionHandlers/attachActionHandlers.cpp:357]: Even if the UE advertises its capability, the network may/may not support the feature.
This flag indicates the result of the negotiation, that takes place to confirm whether the network supports the feature or not.

File: src/mme-app/actionHandlers/attachActionHandlers.cpp (hunk 6 above)
Comment (asreek) [reply src/mme-app/actionHandlers/attachActionHandlers.cpp:569]: Hi Ajay,
We can explore using mem pool managed MsgBuffer objects for encoding q message structures. Will discuss on this during next call.
Thanks,
Anjana

**Comment by thakurajayL** [comment]
retest this please

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**